### DISCUSSION TITLE
Control & Beam & LCP & BarycentricMapping Problems

### DISCUSSION THREAD

**Discussion by jungjae01eng** [sofa-framework/sofa#4154] [Write / Run a simulation] [answered]
Hello, SOFA Community!~

I am new to SOFA and am currently working on my first project! I have spent some time working on this code and have included all my questions in this post. I would **greatly appreciate any help, even if it doesn't cover every question**. Thank you so much! ~

---

# 1. Control & Cable Constraint
## Question 1
- **Can I connect the cable to the actuator using indices?**
- On the actuator, I am attempting to connect the cable to the object's mesh # 81; however, when I do so, I encounter the following error.

### Error Message:
```
[ERROR]   [CableConstraint(CableConstraint)] Indices at index 0 is too large regarding mechanicalState [position] size
```

## Question 2
- Does anyone have access to the code for [TetrahedronHyperelasticityFEMForceField.scn](https://github.com/sofa-framework/sofa/blob/master/examples/Components/forcefield/TetrahedronHyperelasticityFEMForceField.scn)?

----

## Question 3
- Furthermore, when I attempt to use the mouse, I encounter the error below. After online search, I found a page with a similar issue. Unfortunately, it appears that this forum was closed without a definitive answer: ["How to attach a beam with a rigid body"](https://www.sofa-framework.org/community/forum/topic/how-to-attach-a-beam-with-a-rigid-body/) by PengYu from SOFA Forum.
- Regarding this matter, I have absolutely no idea, and wasn't able to find any useful solution online. **I suspect that the problem may lie within the controller class, but I'm uncertain**.

### Error Message:
```
[ERROR]   [MechanicalObject(MousePosition)] Invalid vOp operation 1 (null(V_DERIV),0(0),null(V_DERIV),100)
[ERROR]   [MechanicalObject(MappedMousePosition)] Invalid vOp operation 1 (null(V_DERIV),0(0),null(V_DERIV),100)
```

-----

# 2. Beam _(Solved: [comment](https://github.com/sofa-framework/sofa/discussions/4154#discussioncomment-6973748) by @fredroy )_
## Question 4:
- Is this due to my use of "vec3d"? I am attempting to visualize the deformation of the object. **Is there another way to create a beam in SOFA?** I have tried changing my topology to "rigid3d", but it disrupts the entire topology and leads to a Tetrahedron Error.

### Error Message:
![Screenshot from 2023-09-11 16-50-03](https://github.com/sofa-framework/sofa/assets/144021878/954c6426-4944-422c-b500-b21cfac9dcb4)

---

# 3. LCP
##  Question 5:
- I believe this might be a solver issue; however, I have implemented the solver on the rootNode. I used the GenericConstraintSolver, but I still encounter the same error, even when I use the LCPConstraintSolver. I simply do not fully comprehend why my solver isn't functioning correctly.

### Error Message:
![Screenshot from 2023-09-11 16-29-06](https://github.com/sofa-framework/sofa/assets/144021878/764aa231-eebc-43ad-94f6-def8190ea83f)

---

# 4. BarycentricMapping
## Question 6:
- It seems that most examples involving deformation also utilize BarycentricMapping. I've tried to follow the same approach as others, but I'm unable to resolve this error.

### Error Message:
```
  ValueError: Object type BarycentricMapping<Vec3d,Vec3d> was not created  
The object is in the factory but cannot be created.  
Requested template : None  
Used template      : Vec3d,Vec3d (default)  
Reason(s)          : Data attribute 'input' does not point to a mechanical state of data type 'Vec3d' and none can be found in the parent node context.  
Also tried to create the object with the template 'Vec3d,Rigid3d' but failed for the following reason(s):  
    1. Data attribute 'input' does not point to a mechanical state of data type 'Vec3d' and none can be found in the parent node context.
```

**Comment by fredroy** [comment]
Hello,
Thank you for your really precise description of your problem(s) 😅

First thing (which could fix other problem maybe), for your question #4: the component your are trying to use (`BeamInterpolation`) is contained in the plugin called `BeamAdapter`, which I dont see in the list of plugins you are loading.

**Reply by jungjae01eng** [reply]
Hello, @fredroy~

Thank you so much for your reply!!

Sorry, it ended up being quite lengthy.😅 I have been trying to figure all of this out by myself for the last few months and couldn't find a solution. This is my first time using any kind of simulation, so I have been struggling with understanding how the structure works.

You are right; I forgot to include the `BeamAdapter` plugin, as you mentioned. Nevertheless, I am still facing the same error even after adding the `BeamAdapter` plugin. I just double-checked, and the BeamAdapter file does exist in the plugins folder.

I am a bit curious, though, as to why they are trying to "create the object with the template Rigid3d." Does the mstate have to be Rigid3d to be able to use the beam? I am only trying to apply the beam to my cylinder, which is why I placed this object in softObjectNode (a child of origamiNode). Perhaps I am implementing this incorrectly?


## Error Message:
```
[ERROR]   [SofaPython3::SceneLoader] Unable to completely load the scene from file 'projects/origami/scene.py'.  
Python exception:   
  ValueError: Object type BeamInterpolation<> was not created  
The object is in the factory but cannot be created.  
Requested template : None  
Used template      : None  
Also tried to create the object with the template 'Rigid3d' but failed for the following reason(s):  
    1. No mechanical state with the datatype 'Rigid3d' found in the context node
```

**Reply by fredroy** [reply]
`BeamInterpolation` and `AdaptiveBeamForceFieldAndMass `  (from https://github.com/sofa-framework/BeamAdapter) implement the Kirchhoff  theory and need a set of Rigid (i.e position + orientation), so thats why they need a mechancial object with a "rigid" template

If I understand well your problematic, you basically want to rigidify a soft beam with a (rigid) cable inside right ?
There is a simple example `examples/Demos/rigidifiedSectionsInBeam.scn`

<img src="https://github.com/sofa-framework/sofa/assets/11028016/0f29ec10-fc5e-4b86-b8c1-1bf5dea9dfb4" width="200" />

And you may refer to a similar question here https://github.com/sofa-framework/sofa/discussions/4139

**Reply by jungjae01eng** [reply]
Yeah, you are right. I have just changed the template to rigid3d, and it seems to be working. A few days ago, when I initially changed the template to rigid3d, it triggered some perplexing errors that I couldn't quite comprehend. However, that's no longer the case.

I am interested in designing a deformable cylinder made of silicone that can be controlled.

["Partial Rigidification of a Deformable Object"](https://github.com/sofa-framework/sofa/discussions/4139#discussion-5598660) discussion appears to be addressing the concept of partially rigidifying a deformable object. I will look into this further, but maybe "beam" tool may not be the most suitable option for my needs.

I initially considered using the Beam tool to reduce computational overhead; however, maybe I should just use Tetrahedral FEM for now.
I will look into this and will reach out if I have any more questions regarding this. I truly appreciate your help - thank you, @fredroy!~

**Reply by fredroy** [reply]
What you are describing looks really close to works done with the project/plugin SoftRobots:
https://github.com/SofaDefrost/SoftRobots
You may take a look at this repository and see if it is interesting for you 👍

**Reply by jungjae01eng** [reply]
Thank you!~
 
Yes, I've been playing around with Defrost Team's tutorial, and managed to create deformations of a cylinder. One of their examples, CableGripper, involves cables, which is why I'm currently working on solving Questions 1 to 3.

They used the PullingCable function, which adds cable constraints and defines Barycentric Mapping. I managed to visualize the cable with pullPoint and cableGeometry. However, I'm facing difficulties in actually connecting the cable to the object. They used indices in the cable constraint, which I assume this is the method by which it connects to the object. Nevertheless, despite my efforts to link the cable to the object's mesh at index # 81, I encountered an error (Question 1). Furthermore, when I attempted to use Barycentric Mapping, it gave additional problems (Question 6).

That's why I decided to look into the AttachConstraint function after reading another post from Hugo, but I also encountered an issue (Question 2). :sob:

I will look into the [Defrost page](https://github.com/SofaDefrost/SoftRobots) again - Perhaps I overlooked some crucial details.
Thank you!

**Comment by jungjae01eng** [comment]
Hey @bakpaul,

Sorry, but I'm going to unmarked "Mark as answer" on Frederick's comment because I'm still awaiting some help with the rest of the questions. I am still stuck on questions 1 through 3 and am working on them to figure them out.

**Reply by bakpaul** [reply]
Sorry about that. I was too quick !

**Comment by fredroy** [comment]
As for the Question6, I dont see a Mapping in your snippet, I guess it is somewhere else ?
It could be useful to have a look,  in order to understand why it cannot instantiate it

**Reply by jungjae01eng** [reply]
I apologize for the delayed response. I have resolved all of these questions, so I will close the discussion. I will organize my solution and post it here when I have the time. Thank you very much for your help!

**Comment by jungjae01eng** [comment, marked as answer]
# Solutions
**Question 1:** Yes, and the best example to check out is [SoftGripper.py](https://softroboticstoolkit.com/sofa/tutorial).
**Question 2:** Not sure.
**Question 3:** This can be disregarded based on my knowledge. This error message occurs whenever there is a keyboard interaction, but in my case, you can control the object even with this error without any issues.
**Question 4:** This issue was addressed in another comment.
**Question 5:** I'm uncertain - I ultimately didn't use the LCP solver. However, this might be a useful [link](https://www.sofa-framework.org/community/doc/simulation-principles/constraint/lagrange-constraint/).
**Question 6:** I also didn't end up using Barycentric Mapping; however, I discovered that I had a minor typo issue on my end. For great resources, you can visit this [link](https://www.sofa-framework.org/community/doc/simulation-principles/multi-model-representation/mapping/) and [link](https://project.inria.fr/softrobot/documentation/constraint/cable-actuator/)
